src/api/mod.rs

```diff
@@ -0,0 +1,294 @@
+        Request,
+        Response,
+    },
+    Server,
+};
+use error::ListenerError;
+use futures::Future;
+use hyper::{
+    Method,
+    Uri,
+};
+use lazy_static::lazy_static;
+use mongodb::{
+    options::ClientOptions,
+    Client,
+};
+use pin_project::pin_project;
+use prometheus::{
+    Gauge,
+    HistogramOpts,
+    HistogramVec,
+    IntCounter,
+    IntCounterVec,
+    Opts,
+    Registry,
+};
+use routes::routes;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+use tokio::sync::oneshot;
+use tower::{
+    Layer,
+    Service,
+};
+
+use crate::config::mongo::MongoConfig;
+
+/// The Chronicle API actor
+#[derive(Debug, Serialize, Deserialize, Default)]
+pub struct ChronicleAPI {
+    mongo_config: MongoConfig,
+    #[serde(skip)]
+    server_handle: Option<oneshot::Sender<()>>,
+}
+
+impl ChronicleAPI {
+    pub fn new(mongo_config: MongoConfig) -> Self {
+        Self {
+            mongo_config,
+            server_handle: None,
+        }
+    }
+}
+
+impl PartialEq for ChronicleAPI {
+    fn eq(&self, other: &Self) -> bool {
+        self.mongo_config == other.mongo_config
+    }
+}
+
+impl Clone for ChronicleAPI {
+    fn clone(&self) -> Self {
+        Self {
+            mongo_config: self.mongo_config.clone(),
+            server_handle: None,
+        }
+    }
+}
+
+impl Actor for ChronicleAPI {
+    type Context = Context<Self>;
+
+    fn started(&mut self, _ctx: &mut Self::Context) {
+        register_metrics();
+        let client_opts: ClientOptions = self.mongo_config.clone().into();
+        log::info!("Connecting to MongoDB");
+        let client = Client::with_options(client_opts)
+            .map_err(|e| {
+                log::error!("Unable to create client from options, error: {}", e);
+            })
+            .unwrap();
+        let db = client.database("permanode");
+        let (sender, receiver) = oneshot::channel();
+        log::info!("Starting Axum server");
+        tokio::spawn(async {
+            Server::bind(&([0, 0, 0, 0], 8000).into())
+                .serve(routes(db).into_make_service())
+                .with_graceful_shutdown(shutdown_signal(receiver))
+                .await
+                .unwrap();
+        });
+        self.server_handle = Some(sender);
+    }
+
+    fn stopped(&mut self, _ctx: &mut Self::Context) {
+        if let Some(handle) = self.server_handle.take() {
+            log::info!("Stopping Axum server");
+            handle.send(()).unwrap();
+        }
+        log::info!("Stopping API");
+    }
+}
+
+#[derive(Message)]
+#[rtype(result = "()")]
+pub struct ShutdownAPI;
+
+impl Handler<ShutdownAPI> for ChronicleAPI {
+    type Result = ();
+
+    fn handle(&mut self, _msg: ShutdownAPI, ctx: &mut Self::Context) -> Self::Result {
+        ctx.stop();
+    }
+}
+
+async fn shutdown_signal(recv: oneshot::Receiver<()>) {
+    if let Err(e) = recv.await {
+        log::error!("Error receiving shutdown signal: {}", e);
+    }
+}
+
+fn register_metrics() {
```

Comment:
I would move the whole metrics stuff into its own `metrics` module.

Reply:
Maybe we can even leave out the metrics for now and re-add them at a later point?

Reply:
If the metrics aren't going to be used internally, I doubt they need to exist.

Reply:
I'd say we remove them for now, and them back later using Prometheus. What do you think?

Reply:
Well they are already using prometheus, FYI, and are fully functional.

Reply:
Yeah, sorry. I just realized that.